scipy/spatial/_spherical_voronoi.py

```diff
@@ -161,20 +149,63 @@ def __init__(self, points, radius=1, center=(0, 0, 0), threshold=1e-06):
         self.points = points
         self.radius = radius
         self.center = np.array(center)
+        self.threshold = threshold
 
-        if cKDTree(self.points).query_pairs(threshold * self.radius):
+        if cKDTree(self.points).query_pairs(self.threshold * self.radius):
             raise ValueError("Duplicate generators present.")
 
-        max_discrepancy = sphere_check(self.points,
-                                       self.radius,
-                                       self.center)
-        if max_discrepancy >= threshold * self.radius:
+        radii = np.linalg.norm(self.points - self.center, axis=1)
+        max_discrepancy = np.abs(radii - self.radius).max()
+        if max_discrepancy >= self.threshold * self.radius:
             raise ValueError("Radius inconsistent with generators.")
         self.vertices = None
         self.regions = None
         self._tri = None
         self._calc_vertices_regions()
 
+    def _handle_geodesic_input(self):
+
+        # center the points
+        centered = self.points - self.center
+
+        # calculate an orthogonal transformation using SVD
+        _, _, vh = np.linalg.svd(centered)
+
+        # calculate the north and south poles in this basis
+        poles = [[0, 0, self.radius], [0, 0, -self.radius]] @ vh
+
+        # project points into inverse basis (such that z-components are zero)
+        circle = centered @ vh.T[:, :2]
+
+        # simplicial neighbors are adjacent on the circle
+        angles = np.arctan2(circle[:, 1], circle[:, 0])
+        indices = np.argsort(angles)
+
+        # Voronoi vertices lie halfway between neighboring pairs
+        vertices = centered[indices] + centered[np.roll(indices, 1)]
+        vertices /= np.linalg.norm(vertices, axis=1)[:, np.newaxis]
+        vertices *= self.radius
+
+        # north and south poles are also Voronoi vertices
+        vertices = np.concatenate((vertices, poles))
+
+        # each region contains two vertices from the plane and the north and
+        # south poles
+        invf = np.argsort(indices)
+        invb = np.argsort(np.roll(indices, 1))
```

Comment:
I was looking at this & wondering/ hoping we could avoid  using argsort twice in a row on such a similar array.

I tried to find a pattern and noticed:

```python
for seed in range(100):
    np.random.seed(seed)
    indices = np.arange(15)
    np.random.shuffle(indices)

    invf = np.argsort(indices)
    invb = np.argsort(np.roll(indices, 1))
    print("invf - invb:", invf - invb)   
```

Output always has a bunch of -1 values & (n-1) at `np.argmax(invf)`:


```
invf - invb: [-1 -1 -1 -1 -1 14 -1 -1 -1 -1 -1 -1 -1 -1 -1]
invf - invb: [-1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 14 -1 -1 -1]
invf - invb: [-1 -1 -1 -1 -1 -1 14 -1 -1 -1 -1 -1 -1 -1 -1]
invf - invb: [-1 -1 -1 -1 14 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1]
invf - invb: [-1 -1 -1 -1 -1 -1 -1 -1 -1 -1 14 -1 -1 -1 -1]
invf - invb: [-1 -1 -1 -1 -1 -1 -1 -1 -1 -1 14 -1 -1 -1 -1]
invf - invb: [-1 14 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1]
```

I wonder if that helps us gain efficiency--common offset & one assignment operation?

Reply:
I don't think we can be the simplicity of two argsorts.  Using `line_profiler` on 1000 input points shows that these lines account for about 1.5% of the function time:
```
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   165                                               @profile
   166                                               def _handle_geodesic_input(self):
   167                                           
   168                                                   # center the points
   169         1         12.0     12.0      0.1          centered = self.points - self.center
   170                                           
   171                                                   # calculate an orthogonal transformation using SVD
   172         1      10660.0  10660.0     75.4          _, _, vh = np.linalg.svd(centered)
   173                                           
   174                                                   # calculate the north and south poles in this basis
   175         1         69.0     69.0      0.5          poles = [[0, 0, self.radius], [0, 0, -self.radius]] @ vh
   176                                           
   177                                                   # project points into inverse basis (such that z-components are zero)
   178         1         28.0     28.0      0.2          circle = centered @ vh.T[:, :2]
   179                                           
   180                                                   # simplicial neighbors are adjacent on the circle
   181         1        108.0    108.0      0.8          angles = np.arctan2(circle[:, 1], circle[:, 0])
   182         1         96.0     96.0      0.7          indices = np.argsort(angles)
   183                                           
   184                                                   # Voronoi vertices lie halfway between neighboring pairs
   185         1        227.0    227.0      1.6          vertices = centered[indices] + centered[np.roll(indices, 1)]
   186         1        116.0    116.0      0.8          vertices /= np.linalg.norm(vertices, axis=1)[:, np.newaxis]
   187         1         18.0     18.0      0.1          vertices *= self.radius
   188                                           
   189                                                   # north and south poles are also Voronoi vertices
   190         1         10.0     10.0      0.1          vertices = np.concatenate((vertices, poles))
   191                                           
   192                                                   # each region contains two vertices from the plane and the north and
   193                                                   # south poles
   194         1         63.0     63.0      0.4          invf = np.argsort(indices)
   195         1        136.0    136.0      1.0          invb = np.argsort(np.roll(indices, 1))
   196                                           
   197         1          2.0      2.0      0.0          n = len(self.points)
   198         1          1.0      1.0      0.0          north_pole_index = n
   199         1          2.0      2.0      0.0          south_pole_index = n + 1
   200                                           
   201         1          2.0      2.0      0.0          regions = np.vstack([invf,
   202         1          2.0      2.0      0.0                               [north_pole_index] * n,
   203         1          1.0      1.0      0.0                               invb,
   204         1        316.0    316.0      2.2                               [south_pole_index] * n]).T
   205                                           
   206         1       2238.0   2238.0     15.8          self.regions = [list(region) for region in regions]
   207         1         39.0     39.0      0.3          self.vertices = vertices + self.center
```